### ISSUE TITLE
Correct usage of AntiForgery token in ASP.NET 5 in SPA application?

### ISSUE THREAD

**Issue by artemgrygor** [aspnet/Antiforgery#29]
In previous version of ASP.NET during SPA application the idea of AntiForgey token was following:

* add `@Html.AntiForgeryToken();` on the page
* add `__RequestVerificationToken` to the request
* override AuthorizeAttribute as ValidateJsonAntiForgeryTokenAttribute.

I don't really understand the authorization requirements (is there some good information source?) in ASP.NET 5 but looks like new behavior should be like this:
* add `asp-anti-forgerytaghelper`
* add `__RequestVerificationToken` to the request
* here should be the new requirement.

The question is: how to write this new authorization requirement and remove standard one? 
Could someone give some advice or point me on some example? Thanks

P.S. Looks like I need to create custome middleware and check if it's json to use my way or use standard `await _antiforgery.ValidateRequestAsync(context);`
But still not sure about it.

**Comment by kscott5** [comment]
Then ASP.NET 5 isn't ready? SPA framework namely angularjs using $resource to perform an ajax post to RESTful service. It appears I would need to create some custom pre-filter attribute to validate the antiforgery token. I did notice the dev branch appears to implement a header token to handle this cause. So why wasn't this added to release (RC1)

**Comment by artemgrygor** [comment]
@kscott5 you're right, ASP.NET 5 is only rc1 now. You can't create pre-filter attribute anymore. According to this answer http://stackoverflow.com/a/31465227 
I've found only this implementation on MusicStore https://github.com/aspnet/MusicStore/blob/master/src/MusicStore/Controllers/ShoppingCartController.cs#L62

However this is bad one. So looks like it should be a custome middleware  which reuse a code from MusicStore if it's json or use standard one.

But of course I'm thinking maybe there is something good from asp.net team.

**Comment by davidfowl** [comment]
> So why wasn't this added to release (RC1)

Time machines are hard.

**Comment by rynowak** [comment]
There's a few different approaches and ways to go with this. I've opened a PR with the sample I wrote in parallel to the feature. https://github.com/aspnet/Antiforgery/pull/31

We're certainly all ears when it comes to making this easier in the framework.

This sample is based on Angular and sets up the necessary state to have Angular interoperate ASP.NET Antiforgery. See info here: https://docs.angularjs.org/api/ng/service/$http (ctrl+f XSRF)

1. Some page accessed by the user directly sets the 'request token' in the `XSRF-TOKEN` cookie. In my sample I did this with a simple middleware since the 'landing page' is a static HTML file.
2. Angular will send back the 'request token' in the `X-XSRF-TOKEN` header
3. Configure Antiforgery to look at this header
```
services.AddAntiforgery(options => options.HeaderName = "X-XSRF-TOKEN");
```
4. Your server side code needs to validate the token, either by calling into the service, or using the attribute provided by MVC.

**Comment by kscott5** [comment]
@davidfowl That's hilarious.

@rynowak Unfortunately the AntiforgeryOptions does not include the HeaderName property in RC1 but I believe I can work around this by using the middleware suggestions and avoid the Antiforgery pre-filter attribute.

I see what I come up with.

**Comment by rynowak** [comment]
> Unfortunately the AntiforgeryOptions does not include the HeaderName property in RC1 but I believe I can work around this by using the middleware suggestions and avoid the Antiforgery pre-filter attribute.

@kscott5 - If you want this to work on rc1, I'd suggest replacing the `IAntiforgeryTokenStore` and copying from the rc2 implementation. https://github.com/aspnet/Antiforgery/commit/3280ff6ac55aee20953fbcb7284dcdcb93aa07dd#diff-b6e9c2bfe3926a98f35010c4e094e419R76

Lines 75-88 have the logic you need. The rest of that commit is just renames and error handling.

**Comment by artemgrygor** [comment]
Thanks guys, gonna play with this over the weekend.

**Comment by kscott5** [comment]
I've followed the changes since last year, and impressed with what I've tried.

@rynowak You saved me sometime cause in the past I had to create custom stores for Identity, etc. So you just saved me sometime. Thanks.

**Comment by Eilon** [comment]
@rynowak will your updated features and guidance cover everything discussed here?

**Comment by Eilon** [comment]
@rynowak ping

**Comment by rynowak** [comment]
Yes, I'm not aware of anything else we need to in antiforgery to light up these scenarios.

**Comment by Eilon** [comment]
@rynowak so what MusicStore has right now shows the recommended pattern for antiforgery in a SPA scenario?

**Comment by rynowak** [comment]
yes

**Comment by tinchou** [comment]
Hey guys,

I'm developing a SPA and I'm getting an issue when logging in the user. I tried the approach @rynowak suggested but after logging in I get `Antiforgery token validation failed. The provided antiforgery token was meant for a different claims-based user than the current user.` whenever I do a POST request.

Do you have any suggestion of how I could workaround this?

Thanks!

**Comment by rynowak** [comment]
You will need to refresh the antiforgery token(s) after logging in. The token is tied to the user's identity.

**Comment by tinchou** [comment]
Thanks for your quick reply! I realized I always call `/profile` right after signing in, so I put the token refreshing logic there. It works for my app, but seems hackish.

My main issue is that I'm not using the cookie authentication, but JWT. When the user authenticates I can't call `context.Authentication.SignInAsync` because [it's not implemented](https://github.com/aspnet/Security/blob/22d133cefa20926b25eeb82779343b23a221117f/src/Microsoft.AspNetCore.Authentication.JwtBearer/JwtBearerHandler.cs#L237).

I was thinking maybe manually adding my JWT claims to the `context.User` will make it work since [this is where you get it from](https://github.com/aspnet/Antiforgery/blob/eabe83a72db137afb5cb2d645151233b3f76444b/src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultClaimUidExtractor.cs#L40).

But this won't work with my app since the client side saves the `access_token` in a cookie that is not read by the code when `context.Request.Path == "/"` (it saves the token to avoid signing in each time).

What do you think is a good approach here? I'm probably going to stick with my current solution, but maybe JWT authentication is a scenario we'd want to have better support.

**Comment by rynowak** [comment]
If you're doing bearer auth, you have no reason to use antiforgery.

**Comment by tinchou** [comment]
Jum, that's interesting. I guess I have to search how Bearer Auth protects from CSRF. I haven't seen that mentioned anywhere!

**Comment by rynowak** [comment]
Bearer auth doesn't need CSRF protection because it's not a persistent auth mechanism.

Think of it this way... if a malicious site hosts a form that posts to `example.com`, then `example.com`'s cookies will be included when that form is submitted. This is an automatic behavior of browsers, and it's just how cookies work.

On the other hand if you're using bearer (JWT) using the `Authorize` header then there's nothing automatic. The malicious site doesn't have your JWT, and it won't automatically be provided by the browser - because it's not a cookie.

**Comment by tinchou** [comment]
Oh, I see. So if I store the jwt in a cookie but don't read it in the
server (I only use it as a storage mechanism in the client side) I won't be
subject to XSRF?

**Comment by rynowak** [comment]
>  if I store the jwt in a cookie but don't read it in the
server (I only use it as a storage mechanism in the client side) I won't be
subject to XSRF?

You will absolutely be vulnerable to xsrf if you're using cookies to send it to the server, regardless of what type of token or credential it is. Can you please open a new issue with an example and a more clear description of what you need to accomplish? I think we need more background information to understand what you need.

**Comment by tinchou** [comment]
Sure, I'll see what I can do.

**Comment by danobri** [comment]
@rynowak I'm using your suggested implementation in an Angular 4 SPA, and am wondering how you reccommend handling token expirations?

**Comment by rynowak** [comment]
Our tokens currently don't expire unless the user's session times out. 

I'd suggest getting a new token every time a user logs in

**Comment by danobri** [comment]
From what I can tell, that is precisely the problem - the user's session is timing out, and the token is no longer valid.  The site uses windows authentication so there is no explicit log in.  Here is the error I get:

`CryptographicException: The key {...} was not found in the key ring.
Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingBasedDataProtector.UnprotectCore(Byte[] protectedData, bool allowOperationsOnRevokedKeys, out UnprotectStatus status)
InvalidOperationException: The antiforgery token could not be decrypted.
Microsoft.AspNetCore.Antiforgery.Internal.DefaultAntiforgeryTokenSerializer.Deserialize(string serializedToken)`

Doing some searching on this I discovered that this can happen unless you have explicitly set the application to persist keys, so I followed the instructions [here](https://docs.microsoft.com/en-us/aspnet/core/publishing/iis#data-protection) to create a "registry hive" as the key store, but that hasn't resolved the issue.

**Comment by rynowak** [comment]
Can you please open a new issue for this on aspnet/DataProtection - that sounds like your real problem

**Comment by danobri** [comment]
Sure - but if you are correct that the AntiForgeryToken expires when a user's session expires, seems like I'll still run into this issue regardless of whether the keyring is properly persisted..?

**Comment by rynowak** [comment]
That error doesn't have to do with the token expiring.

**Comment by danobri** [comment]
@rynowak  I was able to resolve the DataProtection error, but I am running into token expirations with users that are testing the application.  I can reproduce the behavior by manually deleting the anti forgery cookies and then trying to make an ajax call.  However, I find if I remove the following, and add the cookie on every response it seems to resolve the issue (at least when I delete the cookie manually).  Any reason the cookie shouldn't be written on every response?

`if (context.Request.Path == "/")`

**Comment by rynowak** [comment]
> I am running into token expirations with users that are testing the application.

Can you describe what you mean by this? Do you mean that the browser has lost track of the cookie? Our token's themselves don't have an expiry, but they are attached to the user's identity.

> Any reason the cookie shouldn't be written on every response?

In general you should only need to do this when you user has to log-in/log-out. I don't see a problem with it though.

**Comment by danobri** [comment]
> Do you mean that the browser has lost track of the cookie? 

Right - when I checked the user's machine the cookie was not present.  I am unable to reproduce myself after leaving a browser session overnight.  Occurred to me that the user may have a plugin that is causing cookies to be cleared, but the user indicated the issue occurs both in IE and Chrome on this particular machine.

> In general you should only need to do this when you user has to log-in/log-out.

The application uses windows authentication, so there is no log in form or explicit log in to hook into.

**Comment by kscott5** [comment]
@artemgrygor I had moved on after my initial assessment but thank you for responding.

**Comment by tarurar** [comment]
@rynowak How am I supposed to refresh the antiforgery token(s) after logging in? Could you please show an example?

**Comment by patricknolan** [comment]
I have an ASPNET Core API \ Angular app. My API needs to support cookies and tokens.

After I login using my service the anti-forgery token returned is not valid as it was created based on a null user. I've  tried setting ClaimsPrincipal after my PasswordSignInAsync and regenerating the anti-forgery token (see below) but that still does not work. Any ideas!

    public virtual async Task<IActionResult> Login([FromBody] AccountLoginModel model)
    {
            var result = await this.SignInManager.PasswordSignInAsync(model.Email, model.Password, isPersistent: model.RememberMe, lockoutOnFailure: false);
            if (!result.Succeeded)
            {
                return this.BadRequest();
            }
            var user = await this.UserManager.FindByEmailAsync(model.Email);

            // Must manually set the HttpContext user claims to those of the logged
            // in user. Otherwise MVC will still include a XSRF token for the "null"
            // user and token validation will fail. (MVC appends the correct token for
            // all subsequent reponses but this isn't good enough for a single page
            // app.)
            var principal = await this.PrincipalFactory.CreateAsync(user);
            this.HttpContext.User = principal;

            // Update XSRF token
            var tokens = this.Antiforgery.GetAndStoreTokens(this.HttpContext);

            return this.Ok();
    }

**Comment by patricknolan** [comment]
I just noticed that this.Antiforgery.GetAndStoreTokens(this.HttpContext); is not changing the value of XSRF-TOKEN. Shouldn't this update based on the fact I have changed the HttpContext.User to be the new principal?

**Comment by MrComic** [comment]
Same Here, It doesn't change the xsrf token if  HttpContext.User Changes After login
here's my description :
https://stackoverflow.com/questions/52290736/xsrf-token-conflicts-with-jwt-token

> I just noticed that this.Antiforgery.GetAndStoreTokens(this.HttpContext); is not changing the value of XSRF-TOKEN. Shouldn't this update based on the fact I have changed the HttpContext.User to be the new principal?
